lgc/disassembler/Disassembler.cpp

```diff
@@ -0,0 +1,506 @@
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  Disassembler.cpp
+ * @brief LGC disassembler library
+ ***********************************************************************************************************************
+ */
+#include "lgc/Disassembler.h"
+#include "llvm/BinaryFormat/MsgPackDocument.h"
+#include "llvm/MC/MCAsmBackend.h"
+#include "llvm/MC/MCAsmInfo.h"
+#include "llvm/MC/MCCodeEmitter.h"
+#include "llvm/MC/MCContext.h"
+#include "llvm/MC/MCDisassembler/MCDisassembler.h"
+#include "llvm/MC/MCInstPrinter.h"
+#include "llvm/MC/MCInstrInfo.h"
+#include "llvm/MC/MCObjectFileInfo.h"
+#include "llvm/MC/MCRegisterInfo.h"
+#include "llvm/MC/MCSectionELF.h"
+#include "llvm/MC/MCStreamer.h"
+#include "llvm/MC/MCSubtargetInfo.h"
+#include "llvm/MC/MCTargetOptions.h"
+#include "llvm/Object/ELFObjectFile.h"
+#include "llvm/Support/AMDGPUMetadata.h"
+#include "llvm/Support/TargetRegistry.h"
+#include "llvm/Support/TargetSelect.h"
+
+using namespace llvm;
+using namespace object;
+
+namespace lgc {
+// Get name of PAL metadata register, or "" if not known
+const char *getPalMetadataRegName(unsigned regNumber);
+} // namespace lgc
+
+namespace {
+
+// Class for the object file disassembler.
+class ObjDisassembler {
+  llvm::MemoryBufferRef m_data;
+  std::unique_ptr<ELFObjectFileBase> m_objFile;
+  llvm::raw_ostream &m_ostream;
+  std::string m_tripleName;
+  const Target *m_target = nullptr;
+  std::unique_ptr<MCRegisterInfo> m_mcRegInfo;
+  std::unique_ptr<MCSubtargetInfo> m_subtargetInfo;
+  std::unique_ptr<MCStreamer> m_streamer;
+  std::unique_ptr<MCDisassembler> m_instDisassembler;
+  MCContext *m_context = nullptr;
+
+public:
+  static void disassembleObject(llvm::MemoryBufferRef data, llvm::raw_ostream &ostream) {
+    ObjDisassembler objDis(data, ostream);
+    objDis.run();
+  }
+
+private:
+  ObjDisassembler(llvm::MemoryBufferRef data, llvm::raw_ostream &ostream) : m_data(data), m_ostream(ostream) {}
+
+  void run();
+  void processSection(ELFSectionRef sectionRef);
+
+  std::vector<object::RelocationRef> relocs;
+
+  void outputData(bool outputting, uint64_t offset, StringRef data,
```

Comment:
nit: I know there's no much consistency in LLVM, but I'd prefer if we used `ArrayRef` for arbitrary data and `StringRef` for actual strings

Reply:
I started trying to do this, but ended up deciding StringRef is better because most of MC deals in it. It is just the LLVM disassembler interface that doesn't.

Reply:
Ack.